agent/src/agentclient.rs

```diff
@@ -0,0 +1,306 @@
+use chrono::naive::NaiveDateTime;
+use k8s_openapi::api::core::v1::Node;
+use kube::api::ListParams;
+use kube::Api;
+use snafu::OptionExt;
+use snafu::ResultExt;
+use std::env;
+use std::process::Command;
+use tokio::time::{sleep, Duration};
+
+use crate::apiclient::{boot_update, get_os_info, list_available, prepare, update};
+use crate::error::{self, Result};
+use apiserver::api::{CreateBottlerocketNodeRequest, UpdateBottlerocketNodeRequest};
+use models::constants::{APISERVER_SERVICE_NAME, APISERVER_SERVICE_PORT, NAMESPACE};
+use models::node::{
+    BottlerocketNode, BottlerocketNodeSelector, BottlerocketNodeState, BottlerocketNodeStatus,
+};
+
+const AGENT_SLEEP_DURATION: Duration = Duration::from_millis(5000);
+
+#[derive(Clone)]
+pub struct BrupopAgent {
+    k8s_client: kube::client::Client,
+}
+
+impl BrupopAgent {
+    pub fn new(k8s_client: kube::client::Client) -> Self {
+        BrupopAgent { k8s_client }
+    }
+
+    pub async fn ensure_node_custom_resource_exists(&self) -> Result<bool> {
+        let associated_bottlerocketnode_name =
+            format!("brn-{}", self.get_node_selector().await?.node_name);
+        let bottlerocket_nodes: Api<BottlerocketNode> =
+            Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+
+        let bottlerocket_nodes_objectlist = bottlerocket_nodes
+            .list(&ListParams::default())
+            .await
+            .context(error::ConvertToObjectList {})?;
+
+        if bottlerocket_nodes_objectlist
+            .iter()
+            .any(|bottlerocket_node_name| {
+                bottlerocket_node_name.metadata.name.as_ref().unwrap()
+                    == &associated_bottlerocketnode_name
+            })
+        {
+            return Ok(true);
+        };
+
+        Ok(false)
+    }
+
+    pub async fn ensure_custom_resource_status_exists(&self) -> Result<bool> {
+        match self.fetch_custom_resource().await?.status {
+            Some(_status) => return Ok(true),
+            None => return Ok(false),
+        };
```

Comment:
You can change this to:

```suggestion
        self.fetch_custom_resource().await?.status.is_some()
```